// @chatui/core/lib/components/SendConfirm/SendConfirm.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SendConfirm = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _Modal = require("../Modal");

var _Flex = require("../Flex");

var _LocaleProvider = require("../LocaleProvider");

var SendConfirm = function SendConfirm(props) {
  var file = props.file,
      onCancel = props.onCancel,
      onSend = props.onSend;

  var _useState = (0, _react.useState)(''),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      img = _useState2[0],
      setImg = _useState2[1];

  var _useLocale = (0, _LocaleProvider.useLocale)('SendConfirm'),
      trans = _useLocale.trans;

  (0, _react.useEffect)(function () {
    if (!file) {
      setImg('');
      return;
    }

    var reader = new FileReader();

    reader.onload = function (e) {
      if (e.target) {
        setImg(e.target.result);
      }
    };

    reader.readAsDataURL(file);
  }, [file]);
  return /*#__PURE__*/_react.default.createElement(_Modal.Modal, {
    className: "SendConfirm",
    title: trans('title'),
    active: !!img,
    vertical: false,
    actions: [{
      label: trans('cancel'),
      onClick: onCancel
    }, {
      label: trans('send'),
      color: 'primary',
      onClick: onSend
    }]
  }, /*#__PURE__*/_react.default.createElement(_Flex.Flex, {
    className: "SendConfirm-inner",
    center: true
  }, /*#__PURE__*/_react.default.createElement("img", {
    src: img,
    alt: ""
  })));
};

exports.SendConfirm = SendConfirm;
