// @chatui/core/es/components/Tabs/Tabs.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import React, { useState, useRef, useEffect } from 'react';
import clsx from 'clsx';
import smoothScroll from '../../utils/smoothScroll';
import useNextId from '../../hooks/useNextId';

var TabItem = function TabItem(props) {
  var active = props.active,
      index = props.index,
      children = props.children,
      onClick = props.onClick,
      others = _objectWithoutProperties(props, ["active", "index", "children", "onClick"]);

  function handleClick(e) {
    onClick(index, e);
  }

  return /*#__PURE__*/React.createElement("div", {
    className: "Tabs-navItem"
  }, /*#__PURE__*/React.createElement("button", _extends({
    className: clsx('Tabs-navLink', {
      active: active
    }),
    type: "button",
    role: "tab",
    "aria-selected": active,
    onClick: handleClick
  }, others), children));
};

var TabsPane = function TabsPane(props) {
  var active = props.active,
      children = props.children,
      others = _objectWithoutProperties(props, ["active", "children"]);

  return /*#__PURE__*/React.createElement("div", _extends({
    className: clsx('Tabs-pane', {
      active: active
    })
  }, others), children);
};

export var Tabs = function Tabs(props) {
  var className = props.className,
      _props$index = props.index,
      oIndex = _props$index === void 0 ? 0 : _props$index,
      scrollable = props.scrollable,
      hideNavIfOnlyOne = props.hideNavIfOnlyOne,
      children = props.children,
      onChange = props.onChange;

  var _useState = useState({}),
      _useState2 = _slicedToArray(_useState, 2),
      pointerStyles = _useState2[0],
      setPointerStyles = _useState2[1];

  var _useState3 = useState(oIndex || 0),
      _useState4 = _slicedToArray(_useState3, 2),
      index = _useState4[0],
      setIndex = _useState4[1];

  var indexRef = useRef(index);
  var navRef = useRef(null);
  var headers = [];
  var contents = [];
  var tabPaneId = useNextId('tabs-');

  function handleIndexChange(idx, e) {
    setIndex(idx);

    if (onChange) {
      onChange(idx, e);
    }
  }

  React.Children.forEach(children, function (item, idx) {
    if (!item) return;
    var active = index === idx;
    var id = "".concat(tabPaneId, "-").concat(idx);
    headers.push( /*#__PURE__*/React.createElement(TabItem, {
      active: active,
      index: idx,
      key: idx,
      onClick: handleIndexChange,
      "aria-controls": id,
      tabIndex: active ? -1 : 0
    }, item.props.label));

    if (item.props.children) {
      contents.push( /*#__PURE__*/React.createElement(TabsPane, {
        active: active,
        key: idx,
        id: id
      }, item.props.children));
    }
  });
  useEffect(function () {
    setIndex(oIndex);
  }, [oIndex]);

  function movePointer() {
    var nav = navRef.current;
    if (!nav) return;
    var currentNav = nav.children[indexRef.current];
    if (!currentNav) return;
    var tab = currentNav.firstChild;
    var _ref = tab,
        tabWidth = _ref.offsetWidth,
        tabOffsetLeft = _ref.offsetLeft;
    setPointerStyles({
      transform: "translateX(".concat(tabOffsetLeft, "px)"),
      width: "".concat(tabWidth, "px")
    });

    if (scrollable) {
      smoothScroll({
        el: nav,
        to: tabOffsetLeft + tabWidth / 2 - nav.offsetWidth / 2,
        x: true
      });
    }
  }

  useEffect(function () {
    var nav = navRef.current;
    var ro;

    if (nav && 'ResizeObserver' in window) {
      ro = new ResizeObserver(movePointer);
      ro.observe(nav);
    }

    return function () {
      if (ro && nav) {
        ro.unobserve(nav);
      }
    };
  }, []);
  useEffect(function () {
    indexRef.current = index;
    movePointer();
  }, [index]);
  var needNav = headers.length > (hideNavIfOnlyOne ? 1 : 0);
  return /*#__PURE__*/React.createElement("div", {
    className: clsx('Tabs', {
      'Tabs--scrollable': scrollable
    }, className)
  }, needNav && /*#__PURE__*/React.createElement("nav", {
    className: "Tabs-nav",
    role: "tablist",
    ref: navRef
  }, headers, /*#__PURE__*/React.createElement("span", {
    className: "Tabs-navPointer",
    style: pointerStyles
  })), /*#__PURE__*/React.createElement("div", {
    className: "Tabs-content"
  }, contents));
};
