// @chatui/core/lib/components/RateActions/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RateActions = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var _IconButton = require("../IconButton");

var GOOD = 'good';
var BAD = 'bad';

var RateActions = function RateActions(props) {
  var goodTitle = props.goodTitle,
      badTitle = props.badTitle,
      onClick = props.onClick;

  var _useState = (0, _react.useState)(''),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      value = _useState2[0],
      setValue = _useState2[1];

  function handleClick(val) {
    if (!value) {
      setValue(val);
      onClick(val);
    }
  }

  function handleGoodClick() {
    handleClick(GOOD);
  }

  function handleBadClick() {
    handleClick(BAD);
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    className: "RateActions"
  }, value !== BAD && /*#__PURE__*/_react.default.createElement(_IconButton.IconButton, {
    className: (0, _clsx.default)('RateBtn', {
      active: value === GOOD
    }),
    title: goodTitle,
    "data-type": "good",
    icon: "thumbs-up",
    onClick: handleGoodClick
  }), value !== GOOD && /*#__PURE__*/_react.default.createElement(_IconButton.IconButton, {
    className: (0, _clsx.default)('RateBtn', {
      active: value === BAD
    }),
    title: badTitle,
    "data-type": "bad",
    icon: "thumbs-down",
    onClick: handleBadClick
  }));
};

exports.RateActions = RateActions;
