// @chatui/core/lib/components/InfiniteScroll/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InfiniteScroll = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireWildcard(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var InfiniteScroll = function InfiniteScroll(props) {
  var className = props.className,
      disabled = props.disabled,
      _props$distance = props.distance,
      distance = _props$distance === void 0 ? 0 : _props$distance,
      children = props.children,
      onLoadMore = props.onLoadMore,
      other = (0, _objectWithoutProperties2.default)(props, ["className", "disabled", "distance", "children", "onLoadMore"]);
  var wrapperRef = (0, _react.useRef)(null);

  function handleScroll() {
    if (disabled) return;
    var el = wrapperRef.current;
    var nearBottom = el.scrollHeight - el.scrollTop - el.clientHeight <= distance;

    if (nearBottom) {
      onLoadMore();
    }
  }

  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    className: (0, _clsx.default)('InfiniteScroll', className),
    role: "feed",
    onScroll: handleScroll,
    ref: wrapperRef
  }, other), children);
};

exports.InfiniteScroll = InfiniteScroll;
