// @chatui/core/lib/components/Button/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Button = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

function composeClass(type) {
  return type && "Btn--".concat(type);
}

var Button = function Button(props) {
  var className = props.className,
      label = props.label,
      color = props.color,
      variant = props.variant,
      size = props.size,
      _props$loading = props.loading,
      loading = _props$loading === void 0 ? false : _props$loading,
      block = props.block,
      _props$disabled = props.disabled,
      disabled = _props$disabled === void 0 ? false : _props$disabled,
      children = props.children,
      onClick = props.onClick,
      other = (0, _objectWithoutProperties2.default)(props, ["className", "label", "color", "variant", "size", "loading", "block", "disabled", "children", "onClick"]);

  function handleClick(e) {
    if (!disabled && !loading && onClick) {
      onClick(e);
    }
  }

  return /*#__PURE__*/_react.default.createElement("button", (0, _extends2.default)({
    className: (0, _clsx.default)('Btn', composeClass(color), composeClass(variant), composeClass(size), {
      'Btn--block': block
    }, className),
    type: "button",
    disabled: disabled,
    onClick: handleClick
  }, other), label || children);
};

exports.Button = Button;
