// @chatui/core/es/components/InfiniteScroll/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import React, { useRef } from 'react';
import clsx from 'clsx';
export var InfiniteScroll = function InfiniteScroll(props) {
  var className = props.className,
      disabled = props.disabled,
      _props$distance = props.distance,
      distance = _props$distance === void 0 ? 0 : _props$distance,
      children = props.children,
      onLoadMore = props.onLoadMore,
      other = _objectWithoutProperties(props, ["className", "disabled", "distance", "children", "onLoadMore"]);

  var wrapperRef = useRef(null);

  function handleScroll() {
    if (disabled) return;
    var el = wrapperRef.current;
    var nearBottom = el.scrollHeight - el.scrollTop - el.clientHeight <= distance;

    if (nearBottom) {
      onLoadMore();
    }
  }

  return /*#__PURE__*/React.createElement("div", _extends({
    className: clsx('InfiniteScroll', className),
    role: "feed",
    onScroll: handleScroll,
    ref: wrapperRef
  }, other), children);
};
