// @chatui/core/lib/components/Message/Message.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var _SystemMessage = require("./SystemMessage");

var _Avatar = require("../Avatar");

var _Time = require("../Time");

var _Typing = require("../Typing");

var Message = function Message(props) {
  var _props$renderMessageC = props.renderMessageContent,
      renderMessageContent = _props$renderMessageC === void 0 ? function () {
    return null;
  } : _props$renderMessageC,
      msg = (0, _objectWithoutProperties2.default)(props, ["renderMessageContent"]);
  var id = msg._id,
      type = msg.type,
      content = msg.content,
      user = msg.user;

  if (type === 'system') {
    return /*#__PURE__*/_react.default.createElement(_SystemMessage.SystemMessage, {
      content: content.text,
      action: content.action,
      key: id
    });
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)('Message', msg.position),
    "data-type": type,
    key: id
  }, msg.hasTime && msg.createdAt && /*#__PURE__*/_react.default.createElement("div", {
    className: "Message-meta"
  }, /*#__PURE__*/_react.default.createElement(_Time.Time, {
    date: msg.createdAt
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "Message-content",
    role: "alert",
    "aria-live": "assertive",
    "aria-atomic": "false"
  }, user && user.avatar && /*#__PURE__*/_react.default.createElement(_Avatar.Avatar, {
    src: user.avatar,
    shape: "square"
  }), type === 'typing' ? /*#__PURE__*/_react.default.createElement(_Typing.Typing, null) : renderMessageContent(msg)));
};

Message.defaultProps = {
  user: {},
  position: 'left',
  hasTime: true
};

var _default = /*#__PURE__*/_react.default.memo(Message);

exports.default = _default;
