// @chatui/core/es/components/Composer/ComposerInput.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";
import React, { useState, useCallback } from 'react';
import clsx from 'clsx';
import { Input } from '../Input';
import { SendConfirm } from '../SendConfirm';
import parseDataTransfer from '../../utils/parseDataTransfer';
export var ComposerInput = function ComposerInput(_ref) {
  var value = _ref.value,
      inputRef = _ref.inputRef,
      invisible = _ref.invisible,
      onImageSend = _ref.onImageSend,
      rest = _objectWithoutProperties(_ref, ["value", "inputRef", "invisible", "onImageSend"]);

  var _useState = useState(null),
      _useState2 = _slicedToArray(_useState, 2),
      pastedImage = _useState2[0],
      setPastedImage = _useState2[1];

  var handlePaste = useCallback(function (e) {
    parseDataTransfer(e, setPastedImage);
  }, []);
  var handleImageCancel = useCallback(function () {
    setPastedImage(null);
  }, []);
  var handleImageSend = useCallback(function () {
    if (onImageSend && pastedImage) {
      Promise.resolve(onImageSend(pastedImage)).then(function () {
        setPastedImage(null);
      });
    }
  }, [onImageSend, pastedImage]);
  return /*#__PURE__*/React.createElement("div", {
    className: clsx({
      'S--invisible': invisible
    })
  }, /*#__PURE__*/React.createElement(Input, _extends({
    className: "Composer-input",
    value: value,
    rows: 1,
    autoSize: true,
    enterKeyHint: "send",
    onPaste: onImageSend ? handlePaste : undefined,
    ref: inputRef
  }, rest)), pastedImage && /*#__PURE__*/React.createElement(SendConfirm, {
    file: pastedImage,
    onCancel: handleImageCancel,
    onSend: handleImageSend
  }));
};
