// @chatui/core/lib/components/QuickReplies/QuickReplies.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _ScrollView = require("../ScrollView/ScrollView");

var _QuickReply = require("./QuickReply");

var QuickReplies = function QuickReplies(props) {
  var items = props.items,
      visible = props.visible,
      onClick = props.onClick,
      onScroll = props.onScroll;
  var scroller = (0, _react.useRef)(null);

  var _useState = (0, _react.useState)(!!onScroll),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      scrollEvent = _useState2[0],
      setScrollEvent = _useState2[1];

  (0, _react.useLayoutEffect)(function () {
    if (scroller.current) {
      setScrollEvent(false);
      scroller.current.scrollTo({
        x: 0,
        y: 0
      });
      setTimeout(function () {
        setScrollEvent(true);
      }, 500);
    }
  }, [items]);
  if (!items.length) return null;
  return /*#__PURE__*/_react.default.createElement(_ScrollView.ScrollView, {
    className: "QuickReplies",
    data: items,
    itemKey: "code",
    ref: scroller,
    "data-visible": visible,
    fullWidth: true,
    onScroll: scrollEvent ? onScroll : undefined,
    renderItem: function renderItem(item, index) {
      return /*#__PURE__*/_react.default.createElement(_QuickReply.QuickReply, {
        item: item,
        index: index,
        onClick: onClick,
        key: item.name
      });
    }
  });
};

QuickReplies.defaultProps = {
  items: [],
  visible: true
};

var _default = /*#__PURE__*/_react.default.memo(QuickReplies);

exports.default = _default;
