// @chatui/core/lib/components/Flex/FlexItem.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FlexItem = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var FlexItem = function FlexItem(props) {
  var className = props.className,
      flex = props.flex,
      alignSelf = props.alignSelf,
      order = props.order,
      children = props.children,
      other = (0, _objectWithoutProperties2.default)(props, ["className", "flex", "alignSelf", "order", "children"]);
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    className: (0, _clsx.default)('FlexItem', className),
    style: {
      flex: flex,
      alignSelf: alignSelf,
      order: order
    }
  }, other), children);
};

exports.FlexItem = FlexItem;
