// @chatui/core/es/components/RateActions/index.js
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import React, { useState } from 'react';
import clsx from 'clsx';
import { IconButton } from '../IconButton';
var GOOD = 'good';
var BAD = 'bad';
export var RateActions = function RateActions(props) {
  var goodTitle = props.goodTitle,
      badTitle = props.badTitle,
      onClick = props.onClick;

  var _useState = useState(''),
      _useState2 = _slicedToArray(_useState, 2),
      value = _useState2[0],
      setValue = _useState2[1];

  function handleClick(val) {
    if (!value) {
      setValue(val);
      onClick(val);
    }
  }

  function handleGoodClick() {
    handleClick(GOOD);
  }

  function handleBadClick() {
    handleClick(BAD);
  }

  return /*#__PURE__*/React.createElement("div", {
    className: "RateActions"
  }, value !== BAD && /*#__PURE__*/React.createElement(IconButton, {
    className: clsx('RateBtn', {
      active: value === GOOD
    }),
    title: goodTitle,
    "data-type": "good",
    icon: "thumbs-up",
    onClick: handleGoodClick
  }), value !== GOOD && /*#__PURE__*/React.createElement(IconButton, {
    className: clsx('RateBtn', {
      active: value === BAD
    }),
    title: badTitle,
    "data-type": "bad",
    icon: "thumbs-down",
    onClick: handleBadClick
  }));
};
