// @chatui/core/es/components/MessageContainer/index.js
import _extends from "@babel/runtime/helpers/esm/extends";
import React, { useEffect, useRef, useImperativeHandle } from 'react';
import { PullToRefresh } from '../PullToRefresh';
import Message from '../Message/Message';
import canUse from '../../utils/canUse';
export var MessageContainer = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var messages = props.messages,
      loadMoreText = props.loadMoreText,
      onRefresh = props.onRefresh,
      onScroll = props.onScroll,
      renderBeforeMessageList = props.renderBeforeMessageList,
      renderMessageContent = props.renderMessageContent;
  var messagesRef = useRef(null);
  var scroller = useRef(null);
  var lastMessage = messages[messages.length - 1];
  useEffect(function () {
    var wrapper = scroller.current.wrapperRef.current;
    var animated = !!wrapper.scrollTop;
    scroller.current.scrollToEnd({
      animated: animated
    });
  }, [lastMessage]);
  useEffect(function () {
    var wrapper = messagesRef.current;
    if (!wrapper) return;
    var willPreventDefault = canUse('passiveListener') ? {
      passive: false
    } : false;
    var needBlur = false;
    var startY = 0;

    function reset() {
      needBlur = false;
      startY = 0;
    }

    function touchStart(e) {
      var _document = document,
          activeElement = _document.activeElement;

      if (activeElement && activeElement.nodeName === 'TEXTAREA') {
        needBlur = true;
        startY = e.touches[0].clientY;
      }
    }

    function touchMove(e) {
      if (needBlur && Math.abs(e.touches[0].clientY - startY) > 20) {
        document.activeElement.blur();
        reset();
      }
    }

    wrapper.addEventListener('touchstart', touchStart);
    wrapper.addEventListener('touchmove', touchMove, willPreventDefault);
    wrapper.addEventListener('touchend', reset);
    wrapper.addEventListener('touchcancel', reset);
    return function () {
      wrapper.removeEventListener('touchstart', touchStart);
      wrapper.removeEventListener('touchmove', touchMove);
      wrapper.removeEventListener('touchend', reset);
      wrapper.removeEventListener('touchcancel', reset);
    };
  }, []);
  useImperativeHandle(ref, function () {
    return {
      ref: messagesRef,
      scrollToEnd: scroller.current.scrollToEnd
    };
  }, []);

  function handleScroll(e) {
    if (onScroll) {
      onScroll(e);
    }
  }

  return /*#__PURE__*/React.createElement("div", {
    className: "MessageContainer",
    ref: messagesRef,
    tabIndex: -1
  }, renderBeforeMessageList && renderBeforeMessageList(), /*#__PURE__*/React.createElement(PullToRefresh, {
    onRefresh: onRefresh,
    onScroll: handleScroll,
    loadMoreText: loadMoreText,
    ref: scroller
  }, /*#__PURE__*/React.createElement("div", {
    className: "MessageList"
  }, messages.map(function (msg) {
    return /*#__PURE__*/React.createElement(Message, _extends({}, msg, {
      renderMessageContent: renderMessageContent,
      key: msg._id
    }));
  }))));
});
