// @chatui/core/es/components/Input/index.js
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";
import _slicedToArray from "@babel/runtime/helpers/esm/slicedToArray";
import _objectWithoutProperties from "@babel/runtime/helpers/esm/objectWithoutProperties";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

import React, { useState, useEffect, useRef, useContext } from 'react';
import clsx from 'clsx';
import { ThemeContext } from '../Form';

function renderCounter() {
  var value = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';
  var maxLength = arguments.length > 1 ? arguments[1] : undefined;
  return maxLength ? /*#__PURE__*/React.createElement("div", {
    className: "Input-counter"
  }, "".concat(value.length, "/").concat(maxLength)) : null;
}

export var Input = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var className = props.className,
      _props$type = props.type,
      type = _props$type === void 0 ? 'text' : _props$type,
      value = props.value,
      placeholder = props.placeholder,
      _props$rows = props.rows,
      oRows = _props$rows === void 0 ? 1 : _props$rows,
      _props$minRows = props.minRows,
      minRows = _props$minRows === void 0 ? oRows : _props$minRows,
      _props$maxRows = props.maxRows,
      maxRows = _props$maxRows === void 0 ? 5 : _props$maxRows,
      maxLength = props.maxLength,
      multiline = props.multiline,
      autoSize = props.autoSize,
      onChange = props.onChange,
      other = _objectWithoutProperties(props, ["className", "type", "value", "placeholder", "rows", "minRows", "maxRows", "maxLength", "multiline", "autoSize", "onChange"]);

  var initialRows = oRows;

  if (initialRows < minRows) {
    initialRows = minRows;
  } else if (initialRows > maxRows) {
    initialRows = maxRows;
  }

  var _useState = useState(initialRows),
      _useState2 = _slicedToArray(_useState, 2),
      rows = _useState2[0],
      setRows = _useState2[1];

  var _useState3 = useState(21),
      _useState4 = _slicedToArray(_useState3, 2),
      lineHeight = _useState4[0],
      setLineHeight = _useState4[1];

  var inputRef = ref || useRef(null);
  var theme = useContext(ThemeContext);
  var isMultiline = multiline || autoSize || oRows > 1;
  var Element = isMultiline ? 'textarea' : 'input';
  var hasCounter = !!maxLength;
  var isLight = theme === 'light';
  useEffect(function () {
    var lhStr = getComputedStyle(inputRef.current, null).lineHeight;
    var lh = Number(lhStr.replace('px', ''));

    if (lh !== lineHeight) {
      setLineHeight(lh);
    }
  }, []);

  function updateRow() {
    if (!autoSize) return;
    var target = inputRef.current;
    var prevRows = target.rows;
    target.rows = minRows;

    if (placeholder) {
      target.placeholder = '';
    } // eslint-disable-next-line no-bitwise


    var currentRows = ~~(target.scrollHeight / lineHeight);

    if (currentRows === prevRows) {
      target.rows = currentRows;
    }

    if (currentRows >= maxRows) {
      target.rows = maxRows;
      target.scrollTop = target.scrollHeight;
    }

    setRows(currentRows < maxRows ? currentRows : maxRows);

    if (placeholder) {
      target.placeholder = placeholder;
    }
  }

  useEffect(function () {
    if (value === '') {
      setRows(initialRows);
    } else {
      updateRow();
    }
  }, [value]);

  function handleChange(e) {
    updateRow();

    if (onChange) {
      var valueFromEvent = e.target.value;
      var shouldTrim = isMultiline && maxLength && valueFromEvent.length > maxLength;
      var val = shouldTrim ? valueFromEvent.substr(0, maxLength) : valueFromEvent;
      onChange(val, e);
    }
  }

  var inputProps = _objectSpread(_objectSpread({}, other), {}, {
    className: clsx('Input', className),
    type: type,
    ref: inputRef,
    rows: rows,
    value: value,
    placeholder: placeholder,
    maxLength: maxLength,
    onChange: handleChange
  });

  if (isLight || hasCounter) {
    return /*#__PURE__*/React.createElement("div", {
      className: clsx('InputWrapper', {
        'is-light': isLight,
        'has-counter': hasCounter
      })
    }, /*#__PURE__*/React.createElement(Element, inputProps), isLight && /*#__PURE__*/React.createElement("div", {
      className: "Input-line"
    }), renderCounter(value, maxLength));
  }

  return /*#__PURE__*/React.createElement(Element, inputProps);
});
