// @chatui/core/lib/components/Notice/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Notice = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var _Icon = require("../Icon");

var _IconButton = require("../IconButton");

var _utils = require("../../utils");

var Notice = function Notice(props) {
  var content = props.content,
      url = props.url,
      _props$hasClose = props.hasClose,
      hasClose = _props$hasClose === void 0 ? true : _props$hasClose,
      onLinkClick = props.onLinkClick,
      onClose = props.onClose; // 展开还是收起状态

  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      collapsed = _useState2[0],
      setCollapsed = _useState2[1]; // 是否需要展开收起


  var _useState3 = (0, _react.useState)(false),
      _useState4 = (0, _slicedToArray2.default)(_useState3, 2),
      hasMore = _useState4[0],
      setHasMore = _useState4[1];

  var contentRef = (0, _react.useRef)(null);

  function handleToggle(e) {
    setCollapsed(!collapsed);
    e.stopPropagation();
  }

  function handleLinkClick() {
    if (url && onLinkClick) {
      onLinkClick(url);
    }
  }

  (0, _react.useEffect)(function () {
    if (contentRef.current && (0, _utils.reflow)(contentRef.current) > 42) {
      setHasMore(true);
      setCollapsed(true);
    }
  }, [content]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "Notice",
    role: "alert",
    "aria-atomic": true,
    "aria-live": "assertive"
  }, /*#__PURE__*/_react.default.createElement(_Icon.Icon, {
    className: "Notice-icon",
    type: "bullhorn"
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "Notice-content",
    role: "link",
    tabIndex: 0,
    onClick: handleLinkClick
  }, /*#__PURE__*/_react.default.createElement("p", {
    className: (0, _clsx.default)({
      collapsed: collapsed
    }),
    ref: contentRef
  }, content), hasMore && /*#__PURE__*/_react.default.createElement("div", {
    className: "Notice-actions"
  }, /*#__PURE__*/_react.default.createElement(_IconButton.IconButton, {
    className: "Notice-more",
    icon: "chevron-down",
    "aria-expanded": !collapsed,
    onClick: handleToggle
  }))), hasClose && /*#__PURE__*/_react.default.createElement(_IconButton.IconButton, {
    className: "Notice-close",
    icon: "close",
    onClick: onClose,
    "aria-label": "\u5173\u95ED\u901A\u77E5"
  }));
};

exports.Notice = Notice;
