// @chatui/core/lib/components/Toast/Toast.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Toast = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _react = _interopRequireWildcard(require("react"));

var _clsx = _interopRequireDefault(require("clsx"));

var _Icon = require("../Icon");

function renderIcon(type) {
  switch (type) {
    case 'success':
      return /*#__PURE__*/_react.default.createElement(_Icon.Icon, {
        type: "check-circle"
      });

    case 'error':
      return /*#__PURE__*/_react.default.createElement(_Icon.Icon, {
        type: "close-circle"
      });

    case 'loading':
      return /*#__PURE__*/_react.default.createElement(_Icon.Icon, {
        type: "spinner",
        spin: true
      });

    default:
      return null;
  }
}

var Toast = function Toast(props) {
  var content = props.content,
      type = props.type,
      duration = props.duration,
      onUnmount = props.onUnmount;

  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      show = _useState2[0],
      setShow = _useState2[1];

  (0, _react.useEffect)(function () {
    setShow(true);

    if (duration !== -1) {
      setTimeout(function () {
        setShow(false);
      }, duration);
      setTimeout(function () {
        if (onUnmount) {
          onUnmount();
        }
      }, duration + 300);
    }
  }, []);

  function handleClick() {
    if (onUnmount) {
      onUnmount();
    }
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _clsx.default)('Toast', {
      show: show
    }),
    "data-type": type,
    role: "alert"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "Toast-content",
    role: "presentation",
    onClick: handleClick
  }, renderIcon(type), /*#__PURE__*/_react.default.createElement("p", {
    className: "Toast-message"
  }, content)));
};

exports.Toast = Toast;
